File: tests/xls2_test.c
/*
 * OpenCReports test
 * Copyright (C) 2019-2025 Zoltán Böszörményi <zboszor@gmail.com>
 * See COPYING.LGPLv3 in the toplevel directory.
 */

#include <stdio.h>

#include <opencreport.h>
#include "test_common.h"

struct rowdata {
	ocrpt_query *q;
	int32_t row;
};

static void test_newrow_cb(opencreport *o, ocrpt_report *r, void *ptr) {
	struct rowdata *rd = ptr;
	ocrpt_query_result *qr;
	int32_t cols;

	qr = ocrpt_query_get_result(rd->q, &cols);

	printf("Row #%d\n", rd->row++);
	print_result_row("a", qr, cols);

	printf("\n");
}

int main(int argc, char **argv) {
	opencreport *o;
	ocrpt_datasource *ds;
	ocrpt_report *r;
	struct rowdata rd = { .row = 0 };

	if (!ocrpt_pandas_initialize()) {
		fprintf(stderr, "Failed to register pandas datasource driver.\n");
		return 0;
	}

	o = ocrpt_init();

	if (!ocrpt_parse_xml(o, "xlsds.xml")) {
		printf("XML parse error\n");
		ocrpt_free(o);
		ocrpt_pandas_deinitialize();
		return 0;
	}

	ds = ocrpt_datasource_get(o, "pandas");
	if (!ds) {
		fprintf(stderr, "Failed to add a pandas datasource.\n");
		ocrpt_free(o);
		ocrpt_pandas_deinitialize();
		return 1;
	}

	r = ocrpt_part_column_new_report(ocrpt_part_row_new_column(ocrpt_part_new_row(ocrpt_part_new(o))));

	rd.q = ocrpt_query_get(o, "a");
	if (!rd.q) {
		fprintf(stderr, "Failed to add a spreadsheet query.\n");
		ocrpt_free(o);
		ocrpt_pandas_deinitialize();
		return 1;
	}

	ocrpt_report_set_main_query(r, rd.q);

	if (!ocrpt_report_add_new_row_cb(r, test_newrow_cb, &rd)) {
		fprintf(stderr, "Failed to add new row callback.\n");
		ocrpt_free(o);
		ocrpt_pandas_deinitialize();
		return  1;
	}

	ocrpt_execute(o);

	ocrpt_free(o);

	ocrpt_pandas_deinitialize();

	return 0;
}

File: tests/xls2_test.php
<?php

require_once 'test_common.php';

$row = 0;

function test_newrow_cb(OpenCReport $o, OpenCReport\Report $r) {
	global $row;
	global $q;

	$qr = $q->get_result();

	if ($row > 0)
		echo PHP_EOL;
	echo "Row #" . $row . PHP_EOL;
	$row++;

	print_result_row("a", $qr);
}

$o = new OpenCReport();
$o->add_search_path(getcwd());

if (!$o->parse_xml("xlsds.xml")) {
	echo "XML parse error" . PHP_EOL;
	exit(0);
}

$ds = $o->datasource_get("pandas");
if (is_null($ds)) {
	echo "Adding pandas datasource failed" . PHP_EOL;
	exit(0);
}

$q = $o->query_get("a");

if (is_null($q)) {
	echo "Adding query for Sheet1 from test.xls failed" . PHP_EOL;
	exit(0);
}

$r = $o->part_new()->row_new()->column_new()->report_new();
$r->set_main_query($q);

$r->add_new_row_cb("test_newrow_cb");

$o->execute();

echo PHP_EOL;
